Sources/BrowserServicesKit/SecureVault/SecureVaultFactory.swift

```diff
@@ -96,5 +104,26 @@ public class SecureVaultFactory {
     internal func makeKeyStoreProvider() -> SecureVaultKeyStoreProvider {
         return DefaultKeyStoreProvider()
     }
+    
+    internal func createAndInitializeEncryptionProviders() throws -> (SecureVaultCryptoProvider, SecureVaultKeyStoreProvider) {
+        let cryptoProvider = makeCryptoProvider()
+        let keystoreProvider = makeKeyStoreProvider()
+        
+        if try keystoreProvider.l1Key() != nil {
+            return (cryptoProvider, keystoreProvider)
+        } else {
+            let l1Key = try cryptoProvider.generateSecretKey()
+            let l2Key = try cryptoProvider.generateSecretKey()
+            let password = try cryptoProvider.generatePassword()
+            let passwordKey = try cryptoProvider.deriveKeyFromPassword(password)
+            let encryptedL2Key = try cryptoProvider.encrypt(l2Key, withKey: passwordKey)
+
+            try keystoreProvider.storeEncryptedL2Key(encryptedL2Key)
+            try keystoreProvider.storeGeneratedPassword(password)
+            try keystoreProvider.storeL1Key(l1Key)
+            
+            return (cryptoProvider, keystoreProvider)
+        }
+    }
```

Comment:
Not super satisfied with this atm. Database migrations only need the crypto and key store providers, and creating the database provider anyway causes a failed migration as it tries to migrate... while already migrating.

Open for feedback on a cleaner way to handle this.

Reply:
Yeah, let's chat about this as that's a big chunk of duplicated code that I think only needs to exist here.